File: week06/23/sum_left_leaves/main.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// Helper function to create a new tree node
struct TreeNode* createNode(int val) {
    struct TreeNode* newNode = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    newNode->val = val;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function prototype for dfs
void dfs(struct TreeNode* node, int isLeft, int* total);

// Function to find the sum of left leaves
int sumOfLeftLeaves(struct TreeNode* root) {
    int total = 0;
    dfs(root, 0, &total); // Start the DFS traversal with root, isLeft = 0 (not a left child)
    return total;
}

// Helper function to perform DFS
void dfs(struct TreeNode* node, int isLeft, int* total) {
    if (!node) return;

    // If it's a leaf node and it's a left child
    if (!node->left && !node->right && isLeft) {
        *total += node->val;
    }

    // Recursively check left and right children
    dfs(node->left, 1, total);  // isLeft = 1 for left child
    dfs(node->right, 0, total); // isLeft = 0 for right child
}

// Driver code to test the function
int main() {
    // Create a binary tree
    //       3
    //      / \
    //     9   20
    //        /  \
    //       15   7
    struct TreeNode* root = createNode(3);
    root->left = createNode(9);
    root->right = createNode(20);
    root->right->left = createNode(15);
    root->right->right = createNode(7);

    // Test the sumOfLeftLeaves function
    int result = sumOfLeftLeaves(root);
    printf("Sum of left leaves: %d\n", result);  // Output should be 24 (9 + 15)

    // Clean up dynamically allocated memory
    free(root->left);
    free(root->right->left);
    free(root->right->right);
    free(root->right);
    free(root);

    return 0;
}

File: week06/23/sum_left_leaves/main.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sumOfLeftLeaves(root):
    total = 0
    def dfs(node, left):
        nonlocal total
        if not node:
            return
        dfs(node.left, True)
        dfs(node.right, False)

        if not node.left and not node.right and left:
            total += node.val
    dfs(root, False)
    return total

# Driver code to test the function
if __name__ == "__main__":
    # Create a binary tree
    #       3
    #      / \
    #     9   20
    #        /  \
    #       15   7
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)

    # Test the sumOfLeftLeaves function
    result = sumOfLeftLeaves(root)
    print("Sum of left leaves:", result)  # Output should be 24 (9 + 15)
